src/cryptonote_protocol/cryptonote_protocol_handler.inl

```diff
@@ -361,6 +362,194 @@ namespace cryptonote
+      {
+        std::list<blobdata> have_tx;
+        std::list<crypto::hash> need_tx_hashes;
+        
+        transaction tx;
+        crypto::hash tx_hash;
+        std::unordered_map<crypto::hash, blobdata> tx_hash_to_blob;        
+        BOOST_FOREACH(auto& tx_blob, arg.b.txs)
+        {
+          if(parse_and_validate_tx_from_blob(tx_blob, tx))
+          {
+            cryptonote::tx_verification_context tvc = AUTO_VAL_INIT(tvc);
+            m_core.handle_incoming_tx(tx_blob, tvc, true, true);
+            if(tvc.m_verifivation_failed)
+            {
+              LOG_PRINT_CCONTEXT_L1("Block verification failed: transaction verification failed, dropping connection");
+              m_p2p->drop_connection(context);
+              m_core.resume_mine();
+              return 1;
+            }
+            
+            if(!tvc.m_added_to_pool) 
+            {
+              get_transaction_hash(tx, tx_hash);              
+              tx_hash_to_blob[tx_hash] = tx_blob;              
+            }
+          }
+          else
+          {
+            LOG_ERROR_CCONTEXT
+            (
+              "sent wrong tx: failed to parse and validate transaction: \r\n"
+              << epee::string_tools::buff_to_hex_nodelimer(tx_blob) 
+              << "\r\n dropping connection"
+            );
+            
+            m_p2p->drop_connection(context);
+            m_core.resume_mine();
+            return 1;
+          }
+        }
+        
+        BOOST_FOREACH(auto& tx_hash, new_block.tx_hashes)
+        {
+          if(m_core.get_pool_transaction(tx_hash, tx))
+          {
+              have_tx.push_back(tx_to_blob(tx));
+          }
+          else
+          {
+            if(tx_hash_to_blob.find(tx_hash) != tx_hash_to_blob.end()) 
+            {
+                have_tx.push_back(tx_hash_to_blob[tx_hash]);
+            }
+            else
+            {
+              need_tx_hashes.push_back(tx_hash); 
+            }
+          }
+        }
+        
+        if(need_tx_hashes.size() > 0) // drats, we don't have everything..
+        {
+          // request non-mempool txs
+          NOTIFY_REQUEST_FLUFFY_MISSING_TX::request missing_tx_req;
+          missing_tx_req.b = arg.b;
+          missing_tx_req.hop = arg.hop;
+          BOOST_FOREACH(auto& tx_hash, need_tx_hashes)
+          {
+            missing_tx_req.missing_tx_hashes.push_back(tx_hash);
+          }
+          
+          post_notify<NOTIFY_REQUEST_FLUFFY_MISSING_TX>(missing_tx_req, context);
+        }
+        else // whoo-hoo we've got em all ..
+        {
+          block_complete_entry b;
+          b.block = arg.b.block;
+          b.txs = have_tx;
+          
+          std::list<block_complete_entry> blocks;
+          blocks.push_back(b);
+          m_core.prepare_handle_incoming_blocks(blocks);
+
+          for(auto tx_blob_it = arg.b.txs.begin(); tx_blob_it!=arg.b.txs.end();tx_blob_it++)
```

Comment:
This seems wrong. It's going through the optional list of txes (likely empty), instead of the one you built.